[fix] Ensure malformed UUID URLs return 404 #682

stktyagi opened the pull request:
## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- N/A I have updated the documentation.

## Reference to Existing Issue

Closes #682 


## Description of Changes
Applied strict UUID regex (allowing dashless) to URL patterns, replacing permissive pattern.
Updated controller API tests to manually build invalid URLs and assert 404, resolving `NoReverseMatch` errors.
Added regression test to verify malformed admin URLs correctly return 404, preventing bug recurrence.

stktyagi committed:
[Fix] Ensure malformed UUID URLs return 404 #682

Applied strict regex across URLs and updated tests which comply while adding regression test for 404 on malformed URLs.

Fixes #682

stktyagi committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

xoaryaa committed:
[Fix] Fix admin URL reversing in malformed admin URL tests #682

Updated the admin test suite to use robust URL resolution, fixing NoReverseMatch errors caused by hardened regex patterns.

Fixes #682

xoaryaa commented:
Hey @stktyagi  I tried pushing directly to `openwisp/openwisp-controller`, but I don’t have permission (403).

I’ve pushed the fix to my fork as a single commit:
- Commit: `8b552e9409f3e63e5659080fbc3726cfe954826f`
- Branch: `fix-1154-admin-urlname` (repo: `xoaryaa/openwisp-controller`)

To bring it into this PR branch:

```bash
git remote add xoaryaa https://github.com/xoaryaa/openwisp-controller.git
git fetch xoaryaa fix-1154-admin-urlname
git cherry-pick 8b552e9409f3e63e5659080fbc3726cfe954826f

stktyagi commented:
> Hey @stktyagi I tried pushing directly to `openwisp/openwisp-controller`, but I don’t have permission (403).
> 
> I’ve pushed the fix to my fork as a single commit:
> 
valid, I'll have a look and bring in your suitable changes

stktyagi committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

stktyagi commented on openwisp_controller/config/tests/test_admin.py:
```diff
@@ -16,7 +17,7 @@
 from reversion.models import Version
 from swapper import load_model
-
+from openwisp_controller.config.models import Device, Template, Vpn
```
fixed in 8a962997e483ca29b87fd11947afdb4f0397547e

stktyagi committed:
[tests] Fix NoReverseMatch in malformed URL tests #682

Updated history URL resolution to use admin_urlname for compatibility across different admin namespaces in the CI environment.

Fixes #682

nemesifier commented on openwisp_controller/config/tests/pytest.py:
```diff
@@ -25,7 +29,7 @@ class TestDeviceConsumer(CreateDeviceMixin):
                         [
                             re_path(
-                                r"^ws/controller/device/(?P<pk>[^/]+)/$",
+                                rf"^ws/controller/device/(?P<pk>{uuid_regex})/$",
```
Django has uuid path converter that we should be able to reuse here:
https://docs.djangoproject.com/en/5.2/topics/http/urls/#path-converters

I am not sure why this was added here in the first place, can't we just reuse the main url router defined in the application?

nemesifier commented on openwisp_controller/config/tests/test_controller.py:
```diff
@@ -273,9 +273,8 @@ def test_device_checksum_requested_signal_is_emitted(self):
-            reverse("controller:device_checksum", args=[pk]), {"key": d.key}
-        )
+        bad_path = f"/controller/device/checksum/{pk}/"
+        response = self.client.get(bad_path, {"key": d.key})
```
This change is hardcoding the path instead of using reverse, which adds maintenance overhead, plus I don't understand what's the real difference with the previous version, it does not look like something we should really do.

nemesifier commented on openwisp_controller/config/tests/test_controller.py:
```diff
@@ -399,9 +397,8 @@ def test_vpn_checksum(self):
-            reverse("controller:vpn_checksum", args=[pk]), {"key": v.key}
-        )
+        bad_path = f"/controller/vpn/checksum/{pk}/"
+        response = self.client.get(bad_path, {"key": v.key})
```
same here

nemesifier commented on openwisp_controller/config/tests/test_controller.py:
```diff
@@ -974,9 +970,8 @@ def test_deactivated_device_report_status(self):
-            reverse("controller:device_report_status", args=[pk]), {"key": d.key}
-        )
+        bad_path = f"/controller/device/report-status/{pk}/"
+        response = self.client.post(bad_path, {"key": d.key})
```
same here

nemesifier commented on openwisp_controller/config/tests/test_controller.py:
```diff
@@ -1091,9 +1086,8 @@ def test_device_update_info_bad_uuid(self):
-            reverse("controller:device_update_info", args=[pk]), params
-        )
+        bad_path = f"/controller/device/update-info/{pk}/"
+        response = self.client.post(bad_path, params)
```
same here

nemesifier commented on openwisp_controller/config/admin.py:
```diff
@@ -178,11 +191,28 @@ def get_urls(self):
+                rf"^(?P<object_id>({uuid_regex}|__fk__))/change/$",
+                self.admin_site.admin_view(self.change_view),
+                name=f"{url_prefix}_change",
+            ),
```
why do we really need to do this?

nemesifier commented on openwisp_controller/config/utils.py:
```diff
@@ -117,22 +120,22 @@ def get_controller_urls(views_module):
         ),
         re_path(
-            "controller/device/report-status/(?P<pk>[^/]+)/$",
+            rf"controller/device/report-status/(?P<pk>{uuid_regex})/$",
```
same here for tha path converters

nemesifier reviewed (changes requested):
I think the main problem is that we are not using:
https://docs.djangoproject.com/en/5.2/topics/http/urls/#path-converters.

See below for more comments.

xoaryaa committed:
Fix malformed UUID admin URLs with custom converter

stktyagi committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

xoaryaa committed:
[Fix] Add basic formatting #682

Added custom uuid converter and converter-based URL patterns.

Fixes #682

stktyagi committed:
[Fix] Follow-up fixes for UUID admin URL converter #682

Added missed cherry picked commit.

Fixes #682

nemesifier committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

nemesifier committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

stktyagi committed:
[fix] Replace regex URLs with custom Path Converters #682

Replaced legacy regex 're_path' with strict Django 'path' converters and
updated tests to manually construct invalid URLs for 404 testing.

Fixes #682

stktyagi committed:
[fix] Fix admin URL test failures in CI #682

Replaced 'admin_urlname' usage with explicit 'admin:' namespaced URLs in test_malformed_admin_urls_return_404.

Fixes #682

stktyagi committed:
[fix] Fix NoReverseMatch in strict admin URL regression test #682

Updated the test to use explicit 'admin:config_device_change' instead of dynamic app_label.

Fixes #682

stktyagi committed:
[fix] Use correct reverse path #682

Used self.app_label to correctly create reversable path

Fixes #682

stktyagi committed:
[fix] Improve test range #682

Restore previous test format but working and functional.

Fixes #682

stktyagi committed:
[fix] Fix websocket route with path converters #682

Prevents WebSocket connection attempts with malformed or garbage-suffixed URLs

Fixes #682

nemesifier commented on openwisp_controller/config/converters.py:
```diff
@@ -0,0 +1,36 @@
+
+    def to_python(self, value: str) -> str:
+        # normalize to dashed, validates both formats
+        return str(uuid.UUID(value))
```
can you write a unit test for this converter please? No models involved, only pure unit test.

nemesifier reviewed (commented):
Looks good @stktyagi, see my comments below.

stktyagi committed:
[fix] Add tests and docstrings for UUID converters #682

Added docstrings to UUIDAnyConverter and UUIDAnyOrFKConverter and unit tests.

Fixes #682

stktyagi committed:
[fix] Remove redundant comments #682

Remove unnecessary comments since docstring was added.

Fixes #682

nemesifier reviewed (commented):
@stktyagi I have a couple of comments below

stktyagi committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

stktyagi committed:
[fix] Moved imports to the top #682

Moved local imports to the top of the file for consistency.

Fixes #682

stktyagi commented on openwisp_controller/config/apps.py:
```diff
@@ -51,6 +52,18 @@ def ready(self, *args, **kwargs):
+        from django.urls import register_converter
+        from django.urls.converters import get_converters
+
+        from .converters import UUIDAnyConverter, UUIDAnyOrFKConverter
```
Oh I missed this while reviewing the cherry pick, my bad!
Resolving

nemesifier commented on openwisp_controller/config/admin.py:
```diff
@@ -166,23 +167,51 @@ def change_view(self, request, object_id, form_url="", extra_context=None):
-        ] + super().get_urls()
+            # strict overrides for the default admin views (single pattern each)
+            path(
+                "<uuid_or_fk:object_id>/history/",
```
Hence this `uuid_or_fk` thing can be removed.

nemesifier reviewed (changes requested):
❌ I did some local testing. I think this patch has several issues:

- is not properly regression testing the real bug.
- tries to solve something which is not a bug
- introduces a lot of unnecessary code for a simple bug fix

See my comments below.

stktyagi committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

stktyagi committed:
[fix] Keep minimal change for bug fix #682

Refactor the code for bug fix strictly with respect to the reported bug.

Fixes #682

stktyagi committed:
[fix] Added tests for converters #682

Added tests for converters in test_converters

Fixes #682

stktyagi committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

nemesifier committed:
Merge branch 'master' into issues/682-NoReverseMatch-error

nemesifier commented on openwisp_controller/config/tests/test_admin.py:
```diff
@@ -1405,6 +1405,16 @@ def test_change_device_malformed_uuid(self):
+            with self.subTest(model=model_name):
+                url = f"/admin/{self.app_label}/{model_name}/download/{malformed_uuid}/"
+                response = self.client.get(url)
+                self.assertNotEqual(response.status_code, 500)
```
This assertion is too weak. `assertNotEqual(status_code, 500)` passes on 302, 200, or anything that isn't 500. The reason the test passes today is that the request gets a 302 (Django admin's `<path:object_id>/` redirects to `/change/`) and never follows it. If you set `follow=True`, you'll see the 500 because `change_view` then calls `reverse` with the long malformed path and we get `NoReverseMatch`.

Please replace this with the exact regression test for #682. Use the URL from the issue, follow redirects, and assert 404:

```python
def test_change_device_404_malformed_url(self):
    url = (
        '/admin/config/device/'
        'de8fa775-1134-47b6-adc5-2da3d0626c72/history/1564/'
        'undefinedadmin/img/icon-deletelink.svg'
    )
    response = self.client.get(url, follow=True)
    self.assertEqual(response.status_code, 404)
```

With the current branch this would fail with 500, which is what we want a regression test to catch.

nemesifier commented on openwisp_controller/config/admin.py:
```diff
@@ -171,8 +171,8 @@ def get_urls(self):
-            re_path(
-                r"^download/(?P<pk>[^/]+)/$",
+            path(
+                "download/<uuid_any:pk>/",
```
The pattern change alone is not enough to fix #682. The actual failure happens a few lines above in `get_extra_context`, where `reverse("{0}_download".format(prefix), args=[pk])` is called before any validation of `pk`. With a strict pattern (whether this custom converter or Django's built-in `uuid`), `reverse` raises `NoReverseMatch` for malformed input and we get a 500.

Fix it where the failure actually happens. Either validate `pk` with `uuid.UUID(str(pk))` and raise `Http404` on `ValueError`, or extend the existing `try/except (ObjectDoesNotExist, ValidationError)` to also catch `NoReverseMatch`.

nemesifier commented on openwisp_controller/config/converters.py:
```diff
@@ -0,0 +1,27 @@
+        r"[0-9a-fA-F]{4}-"
+        r"[0-9a-fA-F]{12}"
+        r")"
+    )
```
Two points on the scope of this converter:

1. The dashless form is only needed for the device facing controller endpoints (devices send the UUID without dashes for backward compatibility). Admin URLs are not called by devices, they are built from a Python UUID instance which always serializes to the dashed form. For the admin URLs in `config/admin.py` we can just use Django's built-in `<uuid:pk>` and skip this custom code.

2. The regex accepts mixed case (`[0-9a-fA-F]`), Django's built-in `uuid` converter only accepts lowercase. If we have a real reason to support uppercase from devices, please leave a comment in the docstring explaining it, otherwise align with Django's behaviour.

nemesifier commented on openwisp_controller/connection/channels/routing.py:
```diff
@@ -1,12 +1,12 @@
-        re_path(
-            r"^ws/controller/device/(?P<pk>[^/]+)/command$",
+        path(
+            "ws/controller/device/<uuid:pk>/command",
```
There's an inconsistency: this route uses the built-in `<uuid:pk>` but `openwisp_controller/config/tests/pytest.py` uses `<uuid_any:pk>`. The test consumer should match production, otherwise tests are exercising a different URL pattern than what we ship.

This websocket route is used by staff via the admin UI, so dashed UUIDs are fine. My preference is to keep `<uuid:pk>` here and align the test to use `<uuid:pk>` too.

nemesifier commented on openwisp_controller/config/apps.py:
```diff
@@ -53,6 +57,11 @@ def ready(self, *args, **kwargs):
+    def register_path_converters(self):
+        converters = get_converters()
+        if "uuid_any" not in converters:
+            register_converter(UUIDAnyConverter, "uuid_any")
```
Minor: `get_converters` is imported from `django.urls.converters`, which is not part of the documented public API. Since `register_converter` only raises a deprecation warning (not an error) when the same name is registered again, the guard is mostly there to silence the warning. If we want to avoid touching internals, we can either call `register_converter(UUIDAnyConverter, "uuid_any")` unconditionally at module top level (the app starts once per process), or keep the guard but accept the dependency on `django.urls.converters.get_converters`. Not a blocker, just flagging.

nemesifier reviewed (changes requested):
❌ I tested this branch locally and the original bug from #682 is still there.

Steps I ran:

1. Checked out this branch.
2. Visited the exact URL from the issue: `/admin/config/device/de8fa775-1134-47b6-adc5-2da3d0626c72/history/1564/undefinedadmin/img/icon-deletelink.svg`.
3. Django admin's `<path:object_id>/` pattern redirects through `.../change/` (301 then 302) and `change_view` is called with `object_id='de8fa775-1134-47b6-adc5-2da3d0626c72/history/1564/undefinedadmin/img/icon-deletelink.svg'`.
4. `get_extra_context()` calls `reverse("{0}_download".format(prefix), args=[pk])` and `NoReverseMatch` is raised again, so the final response is 500.

Result from my reproduction:

```
Final status: 500
Chain: [('.../svg/', 301), ('.../svg/change/', 302)]
```

So this PR makes the URL pattern stricter but does not fix the actual code path that triggers the bug. It only changes where the failure happens. The new admin test passes today only because it does not follow redirects: it gets the 302, asserts `!= 500`, and moves on.

The real fix belongs in `BaseConfigAdmin.get_extra_context` (around line 144 of `config/admin.py`). Either validate `pk` before calling `reverse`, or include `NoReverseMatch` in the existing `try/except` and raise `Http404`. Something like:

```python
if pk:
    try:
        ctx["download_url"] = reverse("{0}_download".format(prefix), args=[pk])
        has_config = True
        if self.model.__name__ == "Device":
            has_config = self.model.objects.get(pk=pk)._has_config()
    except (ObjectDoesNotExist, ValidationError, NoReverseMatch):
        raise Http404()
    else:
        if not has_config:
            ctx["download_url"] = None
```

This is where the original 500 originates, and it solves the issue regardless of whether the URL pattern is strict or permissive.

A few more inline comments below about the test assertion, the custom converter scope, and consistency between `routing.py` and the test consumer.

stktyagi committed:
[fix] Resolved 500 error on malformed custom admin URLs

Gracefully handles NoReverseMatch exceptions on fallback routes to fix the 500 error, reverts admin URLs to native UUIDs, aligns the custom converter with Django's lowercase behavior, and adds strict regression tests.

nemesifier committed:
[chores] Minor improvements
